### PR TITLE
[ LSTM ] Add Unit test for LSTM Layer

### PR DESCRIPTION
**Pull request by jijoongmoon**
- Fix implementation & add backwarding testcase
```
This commits includes:
  . fix implementation ( apply actvation )
  . add one unittest to check functionality of backwarding

**Self evaluation:**
1. Build test:	 [X]Passed [ ]Failed [ ]Skipped
2. Run test:	 [X]Passed [ ]Failed [ ]Skipped
```
- LSTM Unit Test Cases
```
This PR includes,
 . LSTM Unittest Case
 . python code in genModels

Resolves:

**Self evaluation:**
1. Build test:	 [X]Passed [ ]Failed [ ]Skipped
2. Run test:	 [X]Passed [ ]Failed [ ]Skipped
```
- Add return_sequence 
```
This commit includes,
  . impelementation of return_sequence. If it is true, then it
  generate all the output of time iteration and if it false, then
  only last output is available.
  . add 'return_sequence' keyward

Resolves:

**Self evaluation:**
1. Build test:	 [X]Passed [ ]Failed [ ]Skipped
2. Run test:	 [X]Passed [ ]Failed [ ]Skipped
```
-  Add Unit test for LSTM Layer
```
This commit includes,
  . unit test case for lstm layer
  . fixup of potential bug : tanh prime calculation
  .

**Self evaluation:**
1. Build test:	 [X]Passed [ ]Failed [ ]Skipped
2. Run test:	 [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: jijoong.moon <jijoong.moon@samsung.com>
```

### CODE REVIEW

File: nntrainer/layers/lstm.cpp [hunk 1]
```diff
@@ -41,10 +41,17 @@ int LSTMLayer::initialize(Manager &manager) {
...
-  // outut_dim = [ batch, 1, time_iteration, hidden_size ( unit ) ]
+  // if return_sequences == False :
+  //      output_dim = [ batch, 1, 1, hidden_size (unit)]
+  // else:
+  //      output_dim = [ batch, 1, time_iteration, hidden_size ( unit ) ]
   output_dim[0] = input_dim[0];
   output_dim[0].width(unit);
 
+  if (!return_sequences) {
+    output_dim[0].height(1);
```
Comment (kparichay) [inline nntrainer/layers/lstm.cpp:52]: else case missing -> `output_dim[0].height(input_dim[0].height);`?

File: nntrainer/layers/lstm.cpp [hunk 2]
```diff
@@ -153,7 +170,22 @@ void LSTMLayer::forwarding(bool training) {
   Tensor &bias_h =
     weightAt(static_cast<int>(LSTMParams::bias_h)).getVariableRef();
 
-  Tensor &hidden_ = net_hidden[0]->getVariableRef();
+  mem_cell->getVariableRef().setZero();
+  hidden->getVariableRef().setZero();
+  fgio->getVariableRef().setZero();
+
+  if (training) {
+    mem_cell->getGradientRef().setZero();
```
Comment (kparichay) [inline nntrainer/layers/lstm.cpp:178]: IMO this should be moved to `calcGradient()`.

File: nntrainer/layers/lstm.cpp [hunk 3]
```diff
@@ -153,7 +170,22 @@ void LSTMLayer::forwarding(bool training) {
   Tensor &bias_h =
     weightAt(static_cast<int>(LSTMParams::bias_h)).getVariableRef();
 
-  Tensor &hidden_ = net_hidden[0]->getVariableRef();
+  mem_cell->getVariableRef().setZero();
+  hidden->getVariableRef().setZero();
```
Comment (kparichay) [inline nntrainer/layers/lstm.cpp:174]: some of the setZero are redundant here. Add a TODO to update them later?

File: nntrainer/tensor/tensor.cpp [hunk 4]
```diff
@@ -718,10 +718,14 @@ Tensor &Tensor::dot(Tensor const &m, Tensor &result, bool trans, bool trans_m,
     throw exception::not_supported("Error: support only for rank of dot "
                                    "matrix <= 2");
   }
-  if (trans && dim.rank() > 2) {
-    throw exception::not_supported("Error: support only for rank of dot "
-                                   "matrix <= 2 with trans");
-  }
+
+  // Comment out with intension to support the calculation wrt. batch and height
+  // direction of this tensor. It is OK as long as m is 2D
```
Comment (kparichay) [inline nntrainer/tensor/tensor.cpp:723]: in this case, `m` is treated as a 2D matrix. Can we keep the check and let the caller reshape before the call?

File: test/unittest/unittest_nntrainer_models.cpp [hunk 5]
```diff
@@ -1103,12 +1153,16 @@ INSTANTIATE_TEST_CASE_P(
...
+    mkModelTc(addition_resnet_like_validate, "3:1:1:10", 10),
 
     /// #1192 time distribution inference bug
     // mkModelTc(fc_softmax_mse_distribute_validate, "3:1:5:3", 1),
     // mkModelTc(fc_softmax_cross_distribute_validate, "3:1:5:3", 1),
     // mkModelTc(fc_sigmoid_cross_distribute_validate, "3:1:5:3", 1)
+    mkModelTc(lstm_basic, "1:1:1:1", 1),
+    mkModelTc(lstm_return_sequence, "1:1:2:1", 1),
+    mkModelTc(lstm_return_sequence_with_batch, "2:1:2:1", 1),
+    mkModelTc(lstm_return_sequence_with_batch_n, "2:1:2:1", 2)
```
Comment (kparichay) [inline test/unittest/unittest_nntrainer_models.cpp:1165]: lets add testcases later with unit size more than 1 and long time sequences to ensure stability of our LSTM implementations.

File: nntrainer/layers/lstm.cpp [hunk 6]
```diff
@@ -81,27 +88,31 @@ int LSTMLayer::initialize(Manager &manager) {
...
-    std::make_shared<Var_Grad>(output_dim[0], true, true, "LSTM:mem_cell");
-  mem_cell->getVariableRef().setZero();
-  mem_cell->getGradientRef().setZero();
-
   TensorDim d = input_dim[0];
-  d.width(unit * NUM_GATE);
+  d.width(unit);
 
+  mem_cell = std::make_shared<Var_Grad>(d, true, true, "LSTM:mem_cell");
+  hidden = std::make_shared<Var_Grad>(d, true, true, "LSTM:temp_hidden");
```
Comment (kparichay) [inline nntrainer/layers/lstm.cpp:95]: in my understanding, hidden stores the output of the each time slice. If yes, how about rename to `LSTM:output`?

**Review by kparichay — APPROVED** [review]
LGTM other than the minor comments

File: nntrainer/layers/lstm.cpp (hunk 3 above)
Comment (jijoongmoon) [reply nntrainer/layers/lstm.cpp:174]: No. it is necessary for iteration.

Comment (kparichay) [reply nntrainer/layers/lstm.cpp:174]: given line 237L, i dont think setZero is needed.

File: nntrainer/layers/lstm.h [hunk 7]
```diff
@@ -125,17 +122,17 @@ class LSTMLayer : public Layer {
...
 
   /**
-   * @brief     activation function for h_t : default is tanh
+   * @brief     activation function for h_t : default is sigmoid
    */
   ActiFunc acti_func;
 
   /**
-   * @brief     activation type for recurrent : default is sigmoid
+   * @brief     activation type for recurrent : default is tanh
```
Comment (zhoonit) [inline nntrainer/layers/lstm.h:130]: Could you once checkout the terms?
It seems activation is tanh, recurrent_activation is sigmoid.
https://keras.io/api/layers/recurrent_layers/lstm/

File: nntrainer/layers/lstm.cpp (hunk 2 above)
Comment (jijoongmoon) [reply nntrainer/layers/lstm.cpp:178]: done.

File: test/input_gen/recorder.py [hunk 8]
```diff
@@ -139,10 +142,10 @@ def train_step(model, optimizer, loss_fn, initial_input, label, writer_fn, **kwa
 
     # if traversal order is different from nntrainer, this need to be restructured to rely on the order of output.
     for layer in model.layers:
-        # we might need a bit of reordering if output is more than one
+        if isinstance(layer, K.layers.Permute):
+            continue
```
Comment (zhoonit) [inline test/input_gen/recorder.py:146]: Could you check if this line is used? I have made a patch which removes this line, maybe it got into from rebasing?

File: nntrainer/tensor/tensor.cpp (hunk 4 above)
Comment (jijoongmoon) [reply nntrainer/tensor/tensor.cpp:723]: in a some sense, the reshape is kind of redundant. do we really need to? A x B = C  A could be batch, channel, height, width.. it will store.. like batch, channel, width*height. and B is 2D ( height_A , width ) .. I guess it Is good to go
as long as the width is matched.

Comment (kparichay) [reply nntrainer/tensor/tensor.cpp:723]: then how about just give a warning. This is just to ensure that the caller does not make a mistake.

File: test/input_gen/recorder.py (hunk 8 above)
Comment (jijoongmoon) [reply test/input_gen/recorder.py:146]: done.

File: nntrainer/layers/lstm.h (hunk 7 above)
Comment (jijoongmoon) [reply nntrainer/layers/lstm.h:130]: it is properly implemented in code. update the comment.

File: nntrainer/layers/lstm.cpp (hunk 6 above)
Comment (jijoongmoon) [reply nntrainer/layers/lstm.cpp:95]: done

File: test/unittest/unittest_nntrainer_models.cpp (hunk 5 above)
Comment (jijoongmoon) [reply test/unittest/unittest_nntrainer_models.cpp:1165]: yes. I will later pr

**Review by zhoonit — APPROVED** [review]
LGTM

File: nntrainer/layers/lstm.h (hunk 7 above)
Comment (zhoonit) [reply nntrainer/layers/lstm.h:130]: Okay :)

File: nntrainer/tensor/tensor.cpp (hunk 4 above)
Comment (jijoongmoon) [reply nntrainer/tensor/tensor.cpp:723]: done.

File: nntrainer/layers/lstm.cpp (hunk 3 above)
Comment (jijoongmoon) [reply nntrainer/layers/lstm.cpp:174]: This must be zero for next iteration when it starts.

Comment (kparichay) [reply nntrainer/layers/lstm.cpp:174]: its fine for now, this can be done later when we optimize rnn/lstm implementations.
